### ISSUE TITLE
Impossible to disable heartbeats with BlockingConnection 

### ISSUE THREAD

**Issue by darcoli** [pika/pika#965]
The comment in _tune_heartbeat_timeout states: 

>         :param client_value: None to accept server_value; otherwise, an integral
>          number in seconds; 0 (zero) to disable heartbeat.

However when trying to set heartbeat=0 in the ConnectionParameters, the server value is always respected.

Example code to reproduce with default configuration of rabbitmq (server heartbeat = 60s):

```
#!/usr/bin/env python
import pika
import time

credentials = pika.PlainCredentials('test', 'test')

parameters = pika.ConnectionParameters('127.0.0.1',
                                   5672,
                                   '/',
                                   credentials, 
                                   heartbeat=0)

connection = pika.BlockingConnection(parameters)

channel = connection.channel()

channel.queue_declare(queue='test')

channel.basic_publish(exchange='',
                  routing_key='test',
                  body='test msg')
print("test msg sent")


time.sleep(180)

channel.basic_publish(exchange='',
                  routing_key='test',
                  body='test')
print("test msg sent (180s)!")

connection.close()
```

In the sample code, the second message is not sent and the following exception is given:

`pika.exceptions.ConnectionClosed: (-1, "error(32, 'Broken pipe')")`

In RabbitMQ logs, the following is given:
`missed heartbeats from client, timeout: 60s`
This indicates that the heartbeat=0 ConnectionParameter was not negotiated properly.

**Comment by lukebakken** [comment]
Hello, please see the discussion in #636, #666, #874 and the comment made in 3027890081adaa067268aa4839638a32734c263f by @michaelklishin.

Finally, @vitaly-krugl made [this comment](https://github.com/pika/pika/issues/636#issuecomment-366431658) which applies as well. If you `sleep` longer than the heartbeat interval while using `BlockingConnection`, you will see exactly the issue that you report. You should use `BlockingConnection.sleep` or `BlockingConnection.process_data_events`

**Comment by lukebakken** [comment]
Re-opening to address the misleading comment.

**Comment by darcoli** [comment]
The time.sleep() function was used instead of BlockingConnection.sleep() to deliberately incur the issue. 
The bunny client referenced in #874 (mentioned above)
 https://github.com/ruby-amqp/bunny/blob/master/lib/bunny/session.rb#L1172 also provides a way for clients to disable heartbeats. In my opinion, if the amqp specification allows for the heartbeats to be disabled, clients should have a way to do so.

**Comment by lukebakken** [comment]
I'll let @vitaly-krugl and @michaelklishin comment on the decision that was made.

**Comment by vitaly-krugl** [comment]
Per 0.9.1 AMQP under `tune-ok`: 
> short heartbeat
The delay, in seconds, of the connection heartbeat that the client wants. Zero means the client does not want a heartbeat.

I shy away from having pika make an opinionated decision about this. AMQP spec facilitates this, and there is nothing wrong with pika facilitating it as well. It's easy and let the user decide. Pika defaults to accepting the server's proposal via `DEFAULT_HEARTBEAT_TIMEOUT = None`. But, if the user has a need to disable timeout, they should be able to do just that.

Also, see the pika API documentation for setting heartbeats in connection parameters: https://github.com/pika/pika/blob/f29d904b0030079488489af9b8c35401b24fc87c/pika/connection.py#L368-L372

**Comment by vitaly-krugl** [comment]
https://www.rabbitmq.com/heartbeats.html: "Heartbeats can be disabled by setting the timeout interval to 0. This is not a recommended practice.".

I think we should support disabling it with the warning above in pika's heartbeat arg documentation. It used to work like this until about 5 months ago.

**Comment by darcoli** [comment]
@vitaly-krugl I agree although if clients are disabling heartbeats then one would assume that they understand the consequences. I would "protect" the clients by providing sensible defaults (ie by enabling heartbeats or accepting server setting) if nothing is specified explicitly. But in my opinion giving a warning as well is unnecessary. This is analogous to disabling publish confirmations - in which case no warning is given but still there are consequences which the client must be well aware of.

**Comment by vitaly-krugl** [comment]
The present default is to accept server setting. I meant a warning in the documentation - such as "Heartbeats can be disabled by setting the timeout interval to 0. This is not a recommended practice.", not a warning emitted by code. See

> I think we should support disabling it with the warning above in pika's heartbeat arg *documentation\*. It used to work like this until about 5 months ago.

**Comment by darcoli** [comment]
Ah ok then I fully agree - I thought you meant warning emitted by code ;)

**Comment by lukebakken** [comment]
Fixed in #966
